src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+      }
+    }
+    return errors;
+  }
+
+  private validatePostcode(courtAddress: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const errors: string[] = [];
+    const postcodeRequired = isPrimaryAddress || (!isPrimaryAddress && this.addressFieldsNotEmpty(courtAddress));
+    const postcodeEmpty = !courtAddress?.postcode?.trim();
+
+    if (postcodeRequired && postcodeEmpty) {
+      errors.push(this.postcodeMissingError);
+    }
+
+    if (!postcodeEmpty && !this.postcodeIsValidFormat(courtAddress.postcode)) {
+      errors.push(this.invalidPostcodeError);
+    }
+
+    return errors;
+  }
+
+  private validateAddressTypes(primary: DisplayAddress, secondary: DisplayAddress, writeToUsTypeId: number): string[] {
+    return writeToUsTypeId && !!primary.type_id && !!secondary.type_id && // validate only if types selected in both addresses
+    (this.addressFieldsNotEmpty(secondary) || secondary.postcode?.trim()) && // validate only if secondary has some fields entered
+    (primary.type_id !== writeToUsTypeId && secondary.type_id !== writeToUsTypeId) // at least 1 write address should exist
+      ? [this.multipleVisitAddressError]
+      : [];
+  }
+
+  private addressFieldsNotEmpty(courtAddress: DisplayAddress): boolean {
+    return !!courtAddress.address_lines?.trim() || !!courtAddress.address_lines_cy?.trim() ||
+      !!courtAddress.town?.trim() || !!courtAddress.town_cy?.trim();
+  }
+
+  private postcodeIsValidFormat(postcode: string): boolean {
+    const postcodeRegex = new RegExp('([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|' +
+      '(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))' +
+      '\\s?[0-9][A-Za-z]{2})', 'g');
+
+    const match = postcode.match(postcodeRegex);
+    return match?.length === 1 && match[0] === postcode;
+  }
+
+  private getAddressTypesForSelect(addressTypes: AddressType[], isPrimaryAddress: boolean): SelectItem[] {
+    const allAddressTypes = addressTypes.map((at: AddressType) => ({ value: at.id, text: at.name, selected: false }));
+
+    return isPrimaryAddress
+      ? allAddressTypes
+      : allAddressTypes.filter(at => at.text.toLowerCase() !== this.visitOrContactUsAddressType.toLowerCase());
+  }
+
+  private checkErrorResponseForPostcodeErrors(error: AxiosError, addresses: DisplayCourtAddresses):
+    { primaryInvalid: boolean; secondaryInvalid: boolean; errors: string[] } {
+
+    let primaryPostcodeInvalid = false;
+    let secondaryPostcodeInvalid = false;
+    const errors: string[] = [];
+
+    // We expect an array of invalid postcodes in the body of the response
+    if (Array.isArray(error.response.data)) {
+      const invalidPostcodes = error.response.data as string[];
+
+      invalidPostcodes.forEach(invalidPostcode => {
+        if (invalidPostcode === addresses.primary?.postcode) {
+          primaryPostcodeInvalid = true;
+          errors.push(this.primaryAddressPrefix + this.postcodeNotFoundError);
```

Comment:
This is called when a 400 response is returned, but a not found error message would be indicative of a 404 response.

Reply:
The API returns a 400 Bad Request if an attempt is made to set the entire address to one that contains a non-existent postcode (i.e. fails mapit check).

Reply:
Might be one to query further, not too sure if that's great or not. I would assume a 404 for a not found on a postcode (I think mapit postcode check is the only one we do on the API at the moment), and a 400 for a bad request, a 500 if they are down etc

Reply:
The API endpoint is for updating a court entity's address. The postcode validation is just done as part of that. I think a 404 in this context would be for trying to update a non-existent court entity or similar. A 404 response frmo the API will be caught in the front-end and a generic error displayed.